src/main/java/com/algorand/algosdk/builder/transaction/MethodCallTransactionBuilder.java

```diff
@@ -0,0 +1,171 @@
+     * Specify arguments for the ABI method invocation.
+     * 
+     * This will reset the arguments list to what is passed in by the caller.
+     */
+    public T methodArguments(List<Object> arguments) {
+        this.methodArgs = new ArrayList<>(arguments);
+        return (T) this;
+    }
+
+    /**
+     * Specify arguments for the ABI method invocation.
+     * 
+     * This will add the arguments passed in by the caller to the existing list of arguments.
+     */
+    public T addMethodArguments(List<Object> arguments) {
+        this.methodArgs.addAll(arguments);
+        return (T) this;
+    }
+
+    /**
+     * Specify arguments for the ABI method invocation.
+     * 
+     * This will add the argument passed in by the caller to the existing list of arguments.
+     */
+    public T addMethodArgument(Object argument) {
+        this.methodArgs.add(argument);
+        return (T) this;
+    }
+
+    /**
+     * Specify the signer for this method call transaction.
+     */
+    public T signer(TxnSigner signer) {
+        this.signer = signer;
+        return (T) this;
+    }
+
+    @Override
+    public T accounts(List<Address> accounts) {
+        if (accounts != null)
+            this.foreignAccounts = new ArrayList<>(new HashSet<>(accounts));
+        else
+            this.foreignAccounts.clear();
+        return (T) this;
+    }
+
+    @Override
+    public T foreignApps(List<Long> foreignApps) {
+        if (foreignApps != null)
+            this.foreignApps = new ArrayList<>(new HashSet<>(foreignApps));
+        else
+            this.foreignApps.clear();
+        return (T) this;
+    }
+
+    @Override
+    public T foreignAssets(List<Long> foreignAssets) {
+        if (foreignAssets != null)
+            this.foreignAssets = new ArrayList<>(new HashSet<>(foreignAssets));
+        else
+            this.foreignAssets.clear();
+        return (T) this;
+    }
+
+    @Override
+    public T approvalProgram(TEALProgram approvalProgram) {
+        this.approvalProgram = approvalProgram;
+        return (T) this;
+    }
+
+    @Override
+    public T clearStateProgram(TEALProgram clearStateProgram) {
+        this.clearStateProgram = clearStateProgram;
+        return (T) this;
+    }
+
+    @Override
+    public T localStateSchema(StateSchema localStateSchema) {
+        this.localStateSchema = localStateSchema;
+        return (T) this;
+    }
+
+    @Override
+    public T globalStateSchema(StateSchema globalStateSchema) {
+        this.globalStateSchema = globalStateSchema;
+        return (T) this;
+    }
+
+    @Override
+    public T extraPages(Long extraPages) {
+        if (extraPages == null || extraPages < 0 || extraPages > 3) {
+            throw new IllegalArgumentException("extraPages must be an integer between 0 and 3 inclusive");
+        }
+        this.extraPages = extraPages;
+        return (T) this;
+    }
+
+    /**
+     * Build a MethodCallParams object.
+     */
+    public MethodCallParams build() {
```

Comment:
I don't think you should use `build` here, the other `build()` function is intentionally defined as `final`. When building a MethodCall there would be 2 build functions.

You should be able to move the `MethodCallParameters` logic into `applyTo`. This has the added benefit of reusing the fee setting logic directly rather than creating `MethodCallParams` -> applying the txn logic -> then calling into the TxnBuilder again to build the final app call transaction.

That said, I'm not well versed in the app transaction types and usage patterns. So maybe I'm misunderstanding the behavior.

Reply:
Background: right now the flow of object creation is:
* `MethodCallTransactionBuilder` -> `MethodCallParams` -> `List<TransactionWithSigner>` (this list contains 0 or more "argument" transactions, then a final app call transaction)

The current code works because `MethodCallTransactionBuilder` does not inherit from `TransactionBuilder`, so it doesn't have a superclass with `build()` or `applyTo()` methods yet.

I do like this suggestion though, and it _almost_ works, but there are two things that prevent it:
1. The ultimate product of `MethodCallTransactionBuilder` is a `List<TransactionWithSigner>`, whereas `TransactionBuilder.build()` produces a `Transaction`. I suppose more methods could be added that return the additional information to go from a single `Transaction` to `List<TransactionWithSigner>`, but that doesn't seem optimal.
2. Unlike other transaction builders, there is a loss of information between `MethodCallTransactionBuilder` and its final product, `List<TransactionWithSigner>`; information about the method being called is lost, since only a hash of the method signature ends up in the final app call transaction. For this reason, `AtomicTransactionComposer` needs something that retains information about the method and can generate the final list of transactions, which is the role `MethodCallParams` fills right now.

We could get rid of `MethodCallParams`, but then we'd have to pass a `MethodCallTransactionBuilder` directly to `AtomicTransactionComposer.addMethodCall`. It seems a bit odd to me to call that object a builder when the user never calls a `.build()` on it, but I could be persuaded otherwise.

Reply:
As a result of looking into this, I decided to move the method call transaction creation logic from the `AtomicTransactionComposer` to a new method, `MethodCallParams.createTransactions()`, which is more clear about the flow of creation I talked about above.

Reply:
Thanks for these details. Based on my limited understanding of the behavior, the changes look good to me.